lib/elixir/lib/enum.ex

```diff
@@ -1508,6 +1508,10 @@ defmodule Enum do
   @spec join(t, String.t()) :: String.t()
   def join(enumerable, joiner \\ "")
 
+  def join(enumerable, joiner) when is_list(enumerable) and is_binary(joiner) do
+    join_list(enumerable, joiner)
+  end
```

Comment:
Perhaps move this clause after `join(enumerable, "")` because we likely need to optimize only the one where the string is not empty!

Reply:
Got it! Pushed :slightly_smiling_face: